src/OpenLoco/Windows/Vehicle.cpp

```diff
@@ -3269,14 +3276,581 @@ namespace OpenLoco::Ui::Windows::Vehicle
+            const auto& roadData = roadDataArr[roadElement.sequenceIndex()];
+            auto roadOffset2 = Math::Vector::rotate(Map::Pos2(roadData.x, roadData.y), roadElement.unkDirection());
+            auto roadOffset = Map::Pos3(roadOffset2.x, roadOffset2.y, roadData.z);
+            auto roadFirstTile = loc - roadOffset;
+            uint16_t trackAndDirection = roadElement.unkDirection() | (roadElement.roadId() << 3);
+
+            int32_t bestDistance = std::numeric_limits<int32_t>::max();
+            uint16_t bestProgress = 0;
+            const auto moveInfoArr = Map::TrackData::getRoadSubPositon(trackAndDirection);
+            for (const auto& moveInfo : moveInfoArr)
+            {
+                auto potentialLoc = roadFirstTile + moveInfo.loc;
+                auto viewPos = Map::gameToScreen(potentialLoc, viewport.getRotation());
+                auto uiPos = viewport.mapToUi(viewPos);
+                auto distance = Math::Vector::manhattanDistance(uiPos, cursorLoc);
+                if (distance < bestDistance)
+                {
+                    bestDistance = distance;
+                    bestProgress = std::distance(&*moveInfoArr.begin(), &moveInfo);
+                }
+            }
+            return bestProgress;
+        }
+
+        // 0x00478415
+        static std::optional<GameCommands::VehiclePlacementArgs> getRoadAtCursor(const int16_t x, const int16_t y)
+        {
+            static loco_global<int16_t, 0x0113600C> _113600C;
+            static loco_global<int16_t, 0x0113600E> _113600E;
+
+            _113600C = x;
+            _113600E = y;
+            auto [interaction, viewport] = ViewportInteraction::getMapCoordinatesFromPos(x, y, ~ViewportInteraction::InteractionItemFlags::roadAndTram);
+            if (interaction.type != ViewportInteraction::InteractionItem::road)
+            {
+                return {};
+            }
+
+            auto* roadElement = static_cast<Map::RoadElement*>(interaction.object);
+            Map::Pos3 loc(interaction.pos.x, interaction.pos.y, roadElement->baseZ() * 4);
+            auto progress = getRoadProgressAtCursor({ x, y }, *viewport, *roadElement, loc);
+            const auto& roadDataArr = Map::TrackData::getRoadPiece(roadElement->roadId());
+            const auto& roadData = roadDataArr[roadElement->sequenceIndex()];
+            auto roadOffset2 = Math::Vector::rotate(Map::Pos2(roadData.x, roadData.y), roadElement->unkDirection());
+            auto roadOffset = Map::Pos3(roadOffset2.x, roadOffset2.y, roadData.z);
+            auto roadFirstTile = loc - roadOffset;
+            GameCommands::VehiclePlacementArgs placementArgs;
+            placementArgs.pos = roadFirstTile;
+            placementArgs.trackProgress = progress;
+            placementArgs.trackAndDirection = roadElement->unkDirection() | (roadElement->roadId() << 3);
+            return { placementArgs };
```

Comment:
This is a rather dense block of code. Could you leave a few newlines to group things a little? A comment or two might help as well.